target_chains/solana/programs/solana-receiver/src/tests/simulator.rs

```diff
@@ -0,0 +1,163 @@
+use {
+    crate::ID,
+    solana_program::{
+        bpf_loader_upgradeable::{
+            self,
+            UpgradeableLoaderState,
+        },
+        hash::hash,
+        hash::Hash,
+        instruction::{
+            AccountMeta,
+            Instruction,
+        },
+        native_token::LAMPORTS_PER_SOL,
+        pubkey::Pubkey,
+        rent::Rent,
+        stake_history::Epoch,
+        system_instruction,
+        system_program,
+    },
+    solana_program_test::{
+        BanksClient,
+        BanksClientError,
+        ProgramTest,
+        ProgramTestBanksClientExt,
+        read_file,
+    },
+    solana_sdk::{
+        account::Account,
+        signature::{
+            Keypair,
+            Signer,
+        },
+        transaction::Transaction,
+    },
+    std::{
+        mem::size_of,
+        path::Path,
+    },
+};
+
+/// Simulator for the state of the target chain program on Solana. You can run solana transactions against
+/// this struct to test how pyth instructions execute in the Solana runtime.
+pub struct ProgramSimulator {
+    pub program_id:            Pubkey,
+    banks_client:          BanksClient,
+    /// Hash used to submit the last transaction. The hash must be advanced for each new
+    /// transaction; otherwise, replayed transactions in different states can return stale
+    /// results.
+    last_blockhash:        Hash,
+    programdata_id:        Pubkey,
+    pub upgrade_authority: Keypair,
+    pub genesis_keypair:   Keypair,
+}
+
+impl ProgramSimulator {
+
+    /// Deploys the target chain contract as upgradable
+    pub async fn new() -> ProgramSimulator {
+        let mut bpf_data = read_file(
+            std::env::current_dir()
+                .unwrap()
+                .join(Path::new("../../target/deploy/pyth_solana_receiver.so")),
```

Comment:
this is a little janky but it's the only way that @guibescos found to deploy the program as upgradeable (so you can test program upgrades).

Reply:
For this one you don't need to deploy the program as upgradable, I recommend trying something like what I did here https://github.com/pyth-network/program-authority-escrow/blob/main/programs/program-authority-escrow/src/tests/simulator.rs#L50

Reply:
agree it's not needed right now, but we'll want this as we go forward with the target chain contract (because we'll want to test program upgrades via governance).